pommerman/agents/player_agent.py

```diff
@@ -25,20 +82,40 @@ def __init__(self, agent, control):
                 key.A: 3,
                 key.D: 4,
                 key.E: 5,
-                key.Q: 6 # In Pommerman, this will freeze the game.
+                # key.Q: 6 # In Pommerman, this will freeze the game.
             }
         else:
             raise ValueError("Unknown player control '{}'".format(control))
 
     def act(self, obs, action_space):
-        return self._key_input['curr']
+        # Go through the keys and fire for those that needs repetition (because they're held down)
+        for k, state in self._keystate.items():
+            if state.fires():
+                self._q.append(k)
+                state.fire()
+
+        act = 0
+        if self._q:  # Work off the keys that are queued.
+            act = self._key2act.get(self._q.pop(0), 0)
+        # print(act)
+        return act
 
     @staticmethod
     def has_user_input():
         return True
 
     def on_key_press(self, k, mod):
-        self._key_input['curr'] = self._key2act.get(k, 0)
+        # Ignore if we're not handling the key. Avoids "shadowing" ticks in
+        # multiplayer mode.
+        if k in self._key2act:
+            self._keystate[k] = Keystate()
 
     def on_key_release(self, k, mod):
-        self._key_input['curr'] = 0
+        # It could not be in dict if it was pressed when focus was outside the
+        # window but released when the window had focus.
```

Comment:
I don't understand this comment. Can you elaborate please?

Reply:
I was describing how one can get a `key_release` event without having gotten a corresponding `key_press` event before. But this comment is legacy again, in the current incarnation, the check should be there in all cases! Changing it.